include/oneapi/dpl/pstl/glue_algorithm_impl.h

```diff
@@ -1185,6 +1185,35 @@ lexicographical_compare(_ExecutionPolicy&& __exec, _ForwardIterator1 __first1, _
                                                 __last2, oneapi::dpl::__internal::__pstl_less());
 }
 
+// [shift.left]
+
+template <class _ExecutionPolicy, class _ForwardIterator>
+oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy, _ForwardIterator>
+shift_left(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last,
+           typename std::iterator_traits<_ForwardIterator>::difference_type __n)
+{
+    return oneapi::dpl::__internal::__pattern_shift_left(
+        ::std::forward<_ExecutionPolicy>(__exec), __first, __last, __n,
+        oneapi::dpl::__internal::__is_vectorization_preferred<_ExecutionPolicy, _ForwardIterator>(__exec),
+        oneapi::dpl::__internal::__is_parallelization_preferred<_ExecutionPolicy, _ForwardIterator>(__exec));
+}
+
+// [shift.right]
+
+template <class _ExecutionPolicy, class _BidirectionalIterator>
+oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy, _BidirectionalIterator>
+shift_right(_ExecutionPolicy&& __exec, _BidirectionalIterator __first, _BidirectionalIterator __last,
+            typename std::iterator_traits<_BidirectionalIterator>::difference_type __n)
+{
+    using _ReverseIterator = typename std::reverse_iterator<_BidirectionalIterator>;
+    auto __res = oneapi::dpl::__internal::__pattern_shift_left(
+        ::std::forward<_ExecutionPolicy>(__exec), _ReverseIterator(__last), _ReverseIterator(__first), __n,
```

Comment:
If you apply reverse iterator and run with DPC++ backend, it will lead to creation of temporary buffer because reverse iterator will be recognized as an unknown type of iterator. Seems, we need to create a separate pattern for `shift_right`

Reply:
Why "creation of temporary buffer"? Could you please explain it?
Do you mean that _ReverseIterator adapter works  incorrect with sycl iterator because it is not "honest" iterator?

Reply:
I see. It's processed by internal ranges API. However, seems, we still need to make sure it's passed through ranges to device kernel correctly since it can impact performance. I don't think it's a blocker for this patch if unit tests passed of course.

Reply:
Discussed it offline. 
In case of shift_right algo:

1. For host iterators (or pointers, usm in particular) - all is OK.
2. For SYCL iterators shift_right implementation creates a redundant temporary buffer....  It may be addressed later, I think.

Reply:
It can be added to the documentation as a restriction I believe. Does it make sense to add a comment about that here?

Reply:
In any case for shift_right withc SYCL iterator we need  its own pattern... I've added one.

Reply:
ok. Thanks